### PR TITLE
BUG/ENH: Allow users to switch between calendars

### PR DESCRIPTION
**Pull request by FreddieV4**
You previously could not switch between different calendars because the `trading_calendar` parameter was not being passed into run_algo or `TradingAlgorithm`. Once merged, you should be able to pass which calendar you'd like to use via the CLI using `--trading-calendar`.

e.g.
`zipline run -f algo.py --start 2012-01-01 --end 2013-01-01 --trading-calendar LSE`

**Caveats:**
  - `TradingAlgorithm` and `SimulationParameters` both take a `trading_calendar` object as arguments. Rather than going down the rabbit hole of changing up the implementation of both of those classes, I decided to just allow users to first have the capability of using different calendars, and set aside the re-implementation issue as a separate project at a later point. `TradingEnvironment` also takes a `trading_calendar` object
  - You need data that fits the dates associated with the trading calendar you want to use; e.g., if you choose LSE, you need a bundle that has data for every date in the LSE Calendar; the default bundle does not have that (dates such as January 2nd, and July 4th are missing).

### CODE REVIEW

File: zipline/algorithm.py [hunk 1]
```diff
@@ -287,11 +287,16 @@ def __init__(self, *args, **kwargs):
...
             'trading_calendar',
-            get_calendar("NYSE")
+            "NYSE"
         )
 
+        if not calendar:
+            self.trading_calendar = get_calendar("NYSE")
+        else:
+            self.trading_calendar = get_calendar(calendar)
+
```
Comment (FreddieV4) [inline zipline/algorithm.py]: I'm pretty sure that if I'm using `kwargs.pop()` above that I can remove the `if` statement and just assign `self.trading_calendar` to `get_calendar(calendar)`

Comment (richafrank) [reply zipline/algorithm.py]: Yup, you've encapsulated the logic of the conditional statement in the `pop`: if `trading_calendar` is not in the `kwargs` then use `"NYSE"`.

**Comment by FreddieV4** [comment]
@yankees714 could you take a look at this? I was wondering if/where I should have more test coverage now that we're adding/fixing this.

File: zipline/__main__.py [hunk 2]
```diff
@@ -173,6 +174,12 @@ def _(*args, **kwargs):
     " be written to stdout.",
 )
 @click.option(
+    '--trading-calendar',
+    metavar='TRADING-CALENDAR',
+    default='NYSE',
+    help="The calendar you want to use e.g. LSE. NYSE is the default."
```
[the remark this answers is not in the document]
Comment (FreddieV4) [reply zipline/__main__.py]: TIL. Going to make that change right now

File: zipline/utils/run_algo.py [hunk 3]
```diff
@@ -155,15 +156,20 @@ def choose_loader(column):
         env = TradingEnvironment(environ=environ)
         choose_loader = None
 
+    if not trading_calendar:
+        trading_calendar = get_calendar('NYSE')
```
[the remark this answers is not in the document]
Comment (FreddieV4) [reply zipline/utils/run_algo.py]: Both `TradingAlgorithm` and `SimulationParameters` take a `TradingCalendar` object so I added those lines to get the calendar object specifically for `TradingAlgorithm`, IIRC

If anything, I'm thinking I could just remove the handling from `create_simulation_parameters, because I'm just using the same object here for both trading algo and simparams.

Thoughts?

File: zipline/testing/fixtures.py [hunk 4]
```diff
@@ -598,6 +598,8 @@ class WithSimParams(WithTradingEnvironment):
     SIM_PARAMS_START = alias('START_DATE')
     SIM_PARAMS_END = alias('END_DATE')
 
+    SIM_PARAMS_CALENDAR = get_calendar('NYSE')
```
[the remark this answers is not in the document]
Comment (FreddieV4) [reply zipline/testing/fixtures.py]: I noticed that we were only using `cls.trading_calendar` for testing; figured I'd switch it to just use `NYSE`. But now that you've pointed it out, we'd probably not want this change so we can test across whatever calendar is passed in

File: etc/requirements_dev.txt [hunk 5]
```diff
@@ -60,3 +60,6 @@ piprot==0.9.6
 
 # For mocking out requests fetches
 responses==0.4.0
+
+# Debugging
+pdbpp==0.9.1
```
[the remark this answers is not in the document]
Comment (FreddieV4) [reply etc/requirements_dev.txt]: Sure I think that's fair to do

File: zipline/testing/fixtures.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (FreddieV4) [reply zipline/testing/fixtures.py]: Ah I also saw test failures because the default `trading_calendar` is None:

e.g.
```
======================================================================
FAIL: test_schedule_function_time_rule_positionally_misplaced (tests.test_algorithm.TestAlgoScript)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/freddiev4/Documents/quantopian/zipline/tests/test_algorithm.py", line 2272, in test_schedule_function_time_rule_positionally_misplaced
    data_frequency='minute'
  File "/Users/freddiev4/Documents/quantopian/zipline/zipline/utils/factory.py", line 62, in create_simulation_parameters
    trading_calendar=trading_calendar,
  File "/Users/freddiev4/Documents/quantopian/zipline/zipline/finance/trading.py", line 141, in __init__
    "Must pass in trading calendar!"
AssertionError: Must pass in trading calendar!
```

[the remark this answers is not in the document]
Comment (yankees714) [reply zipline/testing/fixtures.py]: I'm not sure that failure is related, since that test is calling `create_simulation_parameters` directly, not using the `self.sim_params` attribute provided by the fixture

File: zipline/utils/run_algo.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (yankees714) [reply zipline/utils/run_algo.py]: Ah, makes sense that we need to pass an actual `TradingCalendar` when constructing the `TradingAlgorithm`. Seems like we're using `create_simulation_parameters` a bunch of places in the tests where we don't pass a calendar, so probably fine to leave this as-is

**Comment by FreddieV4** [comment]
@yankees714 I think this should be good for a 2nd review

File: zipline/__main__.py [hunk 6]
```diff
@@ -6,8 +6,12 @@
 import pandas as pd
 from six import text_type
 
-from zipline.utils.compat import wraps
 from zipline.data import bundles as bundles_module
+from zipline.utils.calendars.calendar_utils import (
+    get_calendar,
+    _default_calendar_factories
```
Comment (yankees714) [inline zipline/__main__.py]: We probably don't want to keep the leading `_` if we're importing this here.

Better yet, what would you think of including a line in `calendar_utils` like
```
default_calendar_names = _default_calendar_factories.keys()
```
and importing that instead?

Comment (FreddieV4) [reply zipline/__main__.py]: Sounds good to me. I also wrapped it with `sorted()`

**Comment by vonpupp** [comment]
Hi,

I am new to zipline. I am trying to use a 24/7 calendar, so I am trying to use the feature of this PR already to do it in a proper way. I am using zipline installed via pip directly from github's egg (`pip install git+git://github.com/quantopian/zipline')

Unfortunately I am not able to use this feature properly, here is what I am doing:

I edited the `extensions.py` as follows:

```
# A bunch of imports removed for the sake of clarity

class TwentyFourSevenCalendar(TradingCalendar):
    """
    Exchange calendar for 24/7 trading.

    Open Time: 12am, UTC
    Close Time: 11:59pm, UTC

    """
    @property
    def name(self):
        return "POLONIEX"

    @property
    def tz(self):
        return timezone("UTC")

    @property
    def open_time(self):
        return time(0, 0)

    @property
    def close_time(self):
        return time(23,59)

    @lazyval
    def day(self):
        return CustomBusinessDay(
            weekmask='Mon Tue Wed Thu Fri Sat Sun',
        )

start_session = pd.Timestamp('2014-01-07', tz='utc')
end_session = pd.Timestamp('2017-11-13', tz='utc')
assets = ['BTC_USD']

register_calendar(
    'TWENTYFOURSEVEN',
    TwentyFourSevenCalendar(
        start=start_session,
        end=end_session
    )
)
register(
    'custom-csvdir-bundle',
    csvdir_equities(["daily"], '/home/av/repos/customers/mdbasset/csvdir'),
    calendar_name='TWENTYFOURSEVEN',
    start_session=start_session,
    end_session=end_session
)
```
As you can see I am using csvdir bundle which has recently been merged. The ingestion process seems to work fine (no errors and I also verified the directory and the sqlite files exist, so I assume the ingestion is ok). I conclude therefore that the 24/7 calendar works properly during ingestion.

When I try to execute with:
```
zipline run -f csvdir_example.py \
    -b custom-csvdir-bundle \
    --start 2016-1-1 \
    --end 2017-1-1 \
    --data-frequency daily \
    --trading-calendar TWENTYFOURSEVEN \
    --capital-base 100000 \
    -o csvdir_example.pickle
```
I get:

```
Error: Invalid value for "--trading-calendar": invalid choice: TWENTYFOURSEVEN. (choose from BMF, CFE, CME, ICE, LSE, NYSE, TSX, us_futures)
```

Could you please help me?

Thanks a lot.

**Comment by richafrank** [comment]
Thanks @vonpupp - I opened a dedicated issue to track this bug: https://github.com/quantopian/zipline/issues/2018
